mesonbuild/dependencies/dev.py

```diff
@@ -32,30 +32,44 @@ def __init__(self, environment, kwargs):
         super().__init__('gtest', environment, 'cpp', kwargs)
         self.main = kwargs.get('main', False)
         self.src_dirs = ['/usr/src/gtest/src', '/usr/src/googletest/googletest/src']
-        self.detect()
+        self.detect(environment, kwargs)
 
-    def detect(self):
+    def detect(self, environment, kwargs):
         gtest_detect = self.clib_compiler.find_library("gtest", self.env, [])
         gtest_main_detect = self.clib_compiler.find_library("gtest_main", self.env, [])
         if gtest_detect and (not self.main or gtest_main_detect):
+            mlog.debug('Found prebuilt gtest')
             self.is_found = True
             self.compile_args = []
             self.link_args = gtest_detect
             if self.main:
                 self.link_args += gtest_main_detect
             self.sources = []
-            self.prebuilt = True
+            self.gtest_lib_type = 'prebuilt'
         elif self.detect_srcdir():
+            mlog.debug('Found gtest sources')
             self.is_found = True
             self.compile_args = ['-I' + d for d in self.src_include_dirs]
             self.link_args = []
             if self.main:
                 self.sources = [self.all_src, self.main_src]
             else:
                 self.sources = [self.all_src]
-            self.prebuilt = False
+            self.gtest_lib_type = 'building self'
         else:
-            self.is_found = False
+            # use pkgconfig as fallback. Pkgconfig support was added in
+            # https://github.com/google/googletest/pull/1197
+            pkgdep = PkgConfigDependency('gtest_main' if self.main else 'gtest', environment, kwargs)
+            if pkgdep.is_found:
+                mlog.debug('Found gtest via pkgconfig')
+                self.is_found = True
+                self.compile_args = pkgdep.compile_args
+                self.link_args = pkgdep.link_args
+                self.version = pkgdep.version
+                self.sources = []
+                self.gtest_lib_type = 'pkgconfig'
```

Comment:
This isn't a good way of writing this (see #2905 and the resulting PR)
I think support for the `method` kwarg should be added, with `pkg-config` a possible value.
I wonder if pkg-config should be the preferred method (since it generally is), then fallback to custom discovery?

Reply:
1. GTest and GMock have never been pkg-config dependency objects, hence if you tried to query its pkgconfig variables, it alweays would have failed.
2. We do something very similar in the python module too, where we query pkgconfig from the environment as a fallback too.
3. I'm not opposed to a `method` argument, the reason I find it slightly kludgey is because people will hardcode `prebuilt` or something and then the fallback is completely disabled.
4. Making pkgconfig the preferred method changes the behaviour oh so slightly, and I'm worried that somewhere, it will break someone's workflow.

Reply:
@jon-turney look at yoru solution, I agree it's the solution to go with.